kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/parquet/ParquetRecordWriterProvider.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.apache.parquet.avro.AvroParquetWriter;
+import org.apache.parquet.hadoop.ParquetFileWriter;
+import org.apache.parquet.hadoop.ParquetWriter;
+import org.apache.parquet.io.OutputFile;
+import org.apache.parquet.io.PositionOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+public class ParquetRecordWriterProvider implements RecordWriterProvider<S3SinkConnectorConfig> {
+  private static final Logger log = LoggerFactory.getLogger(ParquetRecordWriterProvider.class);
+  private static final String EXTENSION = ".parquet";
+  private final S3Storage storage;
+  private final AvroData avroData;
+
+  ParquetRecordWriterProvider(S3Storage storage, AvroData avroData) {
+    this.storage = storage;
+    this.avroData = avroData;
+  }
+
+  @Override
+  public String getExtension() {
+    return storage.conf().getCompressionCodecName().getExtension() + EXTENSION;
+  }
+
+  @Override
+  public RecordWriter getRecordWriter(final S3SinkConnectorConfig conf, final String filename) {
+    return new RecordWriter() {
+      Schema schema = null;
+      Boolean committed = false;
+      ParquetWriter<GenericRecord> writer;
+
+      @Override
+      public void write(SinkRecord record) {
+        if (schema == null) {
+          schema = record.valueSchema();
+          try {
+            log.info("Opening record writer for: {}", filename);
+            final int pageSize = 64 * 1024;
+            org.apache.avro.Schema avroSchema = avroData.fromConnectSchema(schema);
+
+            writer = AvroParquetWriter
+                    .<GenericRecord>builder(new S3ParquetOutputFile(storage, filename))
+                    .withSchema(avroSchema)
+                    .withWriteMode(ParquetFileWriter.Mode.OVERWRITE)
+                    .withDictionaryEncoding(true)
+                    .withCompressionCodec(storage.conf().getCompressionCodecName())
+                    .withPageSize(pageSize)
+                    .build();
+          } catch (IOException e) {
+            throw new ConnectException(e);
+          }
+        }
+        log.trace("Sink record: {}", record.toString());
+        Object value = avroData.fromConnectData(schema, record.value());
+        try {
+          writer.write((GenericRecord) value);
+        } catch (IOException e) {
+          throw new ConnectException(e);
+        }
+      }
+
+      @Override
+      public void close() {
+        if (committed) {
+          return;
+        }
+        commit();
+      }
+
+      @Override
+      public void commit() {
+        try {
+          committed = true;
+          if (writer != null) {
+            writer.close();
+          }
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      }
+    };
+  }
+
+  private static class S3ParquetOutputFile implements OutputFile {
+    private S3Storage storage;
+    private String filename;
```

Comment:
what this `filename` represents. It's not clear without a closer look in the PR. Would you mind editing the initial comment to describe a bit the implementation? In the meantime I'll take a closer look in the next few days. Thanks!

Reply:
Quick response and will update later with the overall description of the implementation. This filename came from `RecordWriterProvider#getRecordWriter(conf, filename)`. Unlike other format, parquet's writer didn't provide `OutputStream` to writer builder directly, instead it provided `OutputFile` and builder would get stream from its method. Here, the implementation just wrapped `s3Storage` and `filename` into this class and used them to create `OutputStream` when method was called, but it could also wrap a created stream instead. What do you think?